File: 144. Binary Tree Preorder Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        if (root == NULL)
            return {};
        stack <TreeNode*> stack;
        stack.push(root);
        vector <int> res {root->val};
        while (!stack.empty())
        {
            while (root != NULL && root->left != NULL)
            {
                stack.push(root->left);
                root = root->left;
                res.push_back(root->val);
            }
            root = stack.top();
            stack.pop();
            if (root->right)
                stack.push(root->right);
            root = root->right;
            if (root != NULL)
                res.push_back(root->val);
        }
        return res;
    }
};

File: 144. Binary Tree Preorder Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        if not root:
            return []
        stack = [root]
        res = [root.val]
        while stack:
            while root and root.left:
                stack += [root.left]
                root = root.left
                res += [root.val]
            root = stack.pop()
            if root and root.right:
                stack += [root.right]
            root = root.right
            if root:
                res += [root.val]
        return res
